File: src/modules/invitation/invitation.controller.ts
import { Controller, Get, Post, Body, Patch, Param, Delete, UseGuards, HttpCode, Req, Query } from '@nestjs/common';
import { InvitationService } from './invitation.service';
import { CreateInvitationDto } from './dto/create-invitation.dto';
import { InvitationResponseDto } from './dto/invitation-response.dto';
import { ApiTags, ApiOperation, ApiResponse, ApiBody, ApiParam, PartialType, ApiBearerAuth, ApiSecurity } from '@nestjs/swagger';
import { AuthGuard } from '@nestjs/passport';

@ApiTags('Invitations')
@ApiBearerAuth()
@UseGuards(AuthGuard('jwt'))
@Controller('invitations')
export class InvitationController {
    constructor(private readonly invitationService: InvitationService) { }

    @Post()
    @ApiSecurity('jwt')
    @ApiOperation({ summary: 'Create a new invitation' })
    @ApiResponse({ status: 400, description: 'Bad request' })
    create(@Body() createInvitationDto: CreateInvitationDto): Promise<InvitationResponseDto> {
        return this.invitationService.create(createInvitationDto);
    }

    @Get()
    @ApiSecurity('jwt')
    @ApiOperation({ summary: 'Get all invitations' })
    @ApiResponse({
        status: 200,
        description: 'List of all invitations',
        type: InvitationResponseDto,
        isArray: true
    })
    async findAll(
        @Req() req: any,
        @Query('page') page?: number,
        @Query('limit') limit?: number,
        @Query('sortBy') sortBy?: string,
        @Query('sortDir') sortDir?: 'asc' | 'desc'
    ): Promise<{ data: InvitationResponseDto[], total: number }> {
        const data = await this.invitationService.findAll(req.user.userId, page, limit, sortBy, sortDir);
        const total = await this.invitationService.count(req.user.userId);
        return { data, total };
    }

    @Get(':id')
    @ApiSecurity('jwt')
    @ApiOperation({ summary: 'Get invitation by ID' })
    @ApiParam({ name: 'id', description: 'Invitation ID' })
    @ApiResponse({
        status: 200,
        description: 'Invitation found',
        type: InvitationResponseDto,
    })
    @ApiResponse({ status: 404, description: 'Invitation not found' })
    findOne(@Param('id') id: string, @Req() req: any): Promise<InvitationResponseDto> {
        return this.invitationService.findOne(id, req.user.userId);
    }

    @Patch(':id')
    @ApiSecurity('jwt')
    @ApiOperation({ summary: 'Update invitation' })
    @ApiParam({ name: 'id', description: 'Invitation ID' })
    @ApiBody({ type: PartialType(CreateInvitationDto) })
    @ApiResponse({
        status: 200,
        description: 'Invitation updated successfully',
    })
    @ApiResponse({ status: 404, description: 'Invitation not found' })
    update(
        @Param('id') id: string,
        @Body() updateInvitationDto: Partial<CreateInvitationDto>,
        @Req() req: any
    ): Promise<InvitationResponseDto> {
        return this.invitationService.update(id, updateInvitationDto, req.user.userId);
    }

    @Delete(':id')
    @ApiSecurity('jwt')
    @HttpCode(204)
    @ApiOperation({ summary: 'Delete invitation' })
    @ApiParam({ name: 'id', description: 'Invitation ID' })
    @ApiResponse({ status: 204, description: 'Invitation deleted successfully' })
    @ApiResponse({ status: 404, description: 'Invitation not found' })
    remove(@Param('id') id: string, @Req() req: any): Promise<void> {
        return this.invitationService.remove(id, req.user.userId);
    }

    @Post(':id/resend')
    @ApiSecurity('jwt')
    @HttpCode(200)
    @ApiOperation({ summary: 'Resend invitation email' })
    @ApiParam({ name: 'id', description: 'Invitation ID' })
    @ApiResponse({
        status: 200,
        description: 'Invitation email resent successfully'
    })
    @ApiResponse({ status: 404, description: 'Invitation not found' })
    resendEmail(@Param('id') id: string, @Req() req: any): Promise<void> {
        return this.invitationService.resendEmail(id, req.user.userId);
    }
}

File: RapidaQuickstart.DotNet/Controllers/InvitationController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using RapidaQuickstart.DotNet.DTOs;
using RapidaQuickstart.DotNet.Services;
using System.Security.Claims;

namespace RapidaQuickstart.DotNet.Controllers
{
    /// <summary>
    /// Invitation management endpoints
    /// </summary>
    [ApiController]
    [Route("api/invitations")]
    public class InvitationController : ControllerBase
    {
        private readonly IInvitationService _invitationService;

        public InvitationController(IInvitationService invitationService)
        {
            _invitationService = invitationService;
        }

        /// <summary>
        /// Create a new invitation
        /// </summary>
        /// <param name="dto">Invitation data</param>
        /// <returns>Created invitation</returns>
        /// <response code="201">Invitation created successfully</response>
        /// <response code="400">Invalid request data</response>
        /// <response code="401">Unauthorized</response>
        [HttpPost]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> CreateInvitation([FromBody] CreateInvitationDto dto)
        {
            try
            {
                var invitedBy = User.FindFirst("userId")?.Value;
                if (string.IsNullOrEmpty(invitedBy))
                {
                    return Unauthorized(new { message = "Unauthorized" });
                }

                var invitation = await _invitationService.CreateInvitationAsync(dto, invitedBy);
                return CreatedAtAction(nameof(GetInvitationById), new { id = invitation.Id }, invitation);
            }
            catch (InvalidOperationException ex)
            {
                return Conflict(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        /// <summary>
        /// Get all invitations with filtering and pagination
        /// </summary>
        /// <param name="filterDto">Filter and pagination parameters</param>
        /// <returns>Paginated list of invitations</returns>
        /// <response code="200">Returns paginated invitations</response>
        /// <response code="401">Unauthorized</response>
        [HttpGet]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> GetAllInvitations([FromQuery] InvitationFilterDto filterDto)
        {
            try
            {
                var (data, total) = await _invitationService.FindAllAsync(filterDto);
                return Ok(new
                {
                    data,
                    total,
                    page = filterDto.Page,
                    limit = filterDto.Limit,
                    totalPages = (int)Math.Ceiling((double)total / filterDto.Limit)
                });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        /// <summary>
        /// Get invitation by ID
        /// </summary>
        /// <param name="id">Invitation ID</param>
        /// <returns>Invitation</returns>
        /// <response code="200">Returns invitation</response>
        /// <response code="404">Invitation not found</response>
        /// <response code="401">Unauthorized</response>
        [HttpGet("{id}")]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> GetInvitationById(string id)
        {
            try
            {
                var invitation = await _invitationService.FindInvitationByIdAsync(id);
                if (invitation == null)
                {
                    return NotFound(new { message = "Invitation not found" });
                }
                return Ok(invitation);
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }


        /// <summary>
        /// Update invitation
        /// </summary>
        /// <param name="id">Invitation ID</param>
        /// <param name="dto">Updated invitation data</param>
        /// <returns>Updated invitation</returns>
        /// <response code="200">Invitation updated successfully</response>
        /// <response code="404">Invitation not found</response>
        /// <response code="401">Unauthorized</response>
        /// <response code="400">Invalid request data</response>
        [HttpPatch("{id}")]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> UpdateInvitation(string id, [FromBody] UpdateInvitationDto dto)
        {
            try
            {
                var invitation = await _invitationService.UpdateInvitationAsync(id, dto);
                return Ok(invitation);
            }
            catch (InvalidOperationException ex)
            {
                return NotFound(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        /// <summary>
        /// Delete invitation
        /// </summary>
        /// <param name="id">Invitation ID</param>
        /// <returns>Success message</returns>
        /// <response code="200">Invitation deleted successfully</response>
        /// <response code="404">Invitation not found</response>
        /// <response code="401">Unauthorized</response>
        [HttpDelete("{id}")]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> DeleteInvitation(string id)
        {
            try
            {
                await _invitationService.DeleteInvitationAsync(id);
                return Ok(new { message = "Invitation deleted successfully" });
            }
            catch (InvalidOperationException ex)
            {
                return NotFound(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }


        /// <summary>
        /// Resend invitation email
        /// </summary>
        /// <param name="id">Invitation ID</param>
        /// <param name="dto">Resend invitation data</param>
        /// <returns>Success message</returns>
        /// <response code="200">Invitation email sent successfully</response>
        /// <response code="404">Invitation not found</response>
        /// <response code="401">Unauthorized</response>
        /// <response code="400">Invalid request data</response>
        [HttpPost("{id}/resend")]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> ResendInvitation(string id, [FromBody] ResendInvitationDto dto)
        {
            try
            {
                await _invitationService.ResendInvitationAsync(id, dto);
                return Ok(new { message = "Invitation email sent successfully" });
            }
            catch (InvalidOperationException ex)
            {
                return BadRequest(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

    }
}
